File: backend/src/alpha_harness/api/catalog.py
"""Data Explorer: syncing the catalog and querying it."""

from __future__ import annotations

from typing import Annotated, Any

from fastapi import APIRouter, Body, Depends, HTTPException, Query
from pydantic import BaseModel

from ..brain.settings_schema import valid_values
from ..catalog.pyramids import pyramid_grid
from ..catalog.queries import FieldFilter, Tuple4
from ..catalog.sync import SyncTarget, serialise_run
from .deps import State

router = APIRouter(prefix="/api/catalog", tags=["catalog"])


def scope(
    region: Annotated[str, Query(description="e.g. USA, EUR, GLB")],
    delay: Annotated[int, Query(ge=0, le=1)],
    universe: Annotated[str, Query(description="e.g. TOP3000")],
    instrument_type: Annotated[str, Query(alias="instrumentType")] = "EQUITY",
) -> Tuple4:
    """The (instrumentType, region, delay, universe) scope every query needs."""
    return Tuple4(instrument_type=instrument_type, region=region, delay=delay, universe=universe)


Scope = Annotated[Tuple4, Depends(scope)]


# --- syncing --------------------------------------------------------------


class SyncRequest(BaseModel):
    region: str
    delay: int
    universe: str
    instrument_type: str = "EQUITY"
    restart: bool = False


@router.post("/sync")
async def start_sync(payload: SyncRequest, state: State) -> dict[str, Any]:
    """Download one scope into the local catalog.

    Runs in the background; progress arrives over the WebSocket ``sync`` topic.
    ``restart`` wipes the scope first.
    """
    target = SyncTarget(
        instrument_type=payload.instrument_type,
        region=payload.region,
        delay=payload.delay,
        universe=payload.universe,
    )
    run = await state.sync.start(target, restart=payload.restart)
    return serialise_run(run)


async def _markets(state: State) -> list[SyncTarget]:
    """Every EQUITY market the account can simulate, from BRAIN's own settings schema.

    A new region or universe is picked up without a code change.
    """
    schema = await state.auth.cached_settings_schema() or await state.auth.refresh_metadata()
    # ponytail: EQUITY only, the one instrument type the Data Explorer offers.
    base: dict[str, Any] = {"instrumentType": "EQUITY"}
    return [
        SyncTarget(instrument_type="EQUITY", region=region, delay=int(delay), universe=universe)
        for region in valid_values(schema, "region", base)
        # ponytail: ALL left out for now. BRAIN pages it 50 fields at a time, ~1,700 requests
        # at 30 a minute; drop this line (and the frontend's HIDDEN_REGIONS) to bring it back.
        if region != "ALL"
        for delay in valid_values(schema, "delay", {**base, "region": region})
        for universe in valid_values(schema, "universe", {**base, "region": region, "delay": delay})
    ]


@router.get("/markets")
async def markets(state: State) -> list[dict[str, Any]]:
    """Every market a full sync covers: what the Data Explorer's sync matrix draws."""
    return [
        {
            "instrumentType": t.instrument_type,
            "region": t.region,
            "delay": t.delay,
            "universe": t.universe,
        }
        for t in await _markets(state)
    ]


@router.post("/sync-all")
async def start_sync_all(state: State) -> dict[str, Any]:
    """Download every market BRAIN offers: all fields first, then dataset details.

    Runs in the background; progress, including each market's state, arrives over the
    WebSocket ``sync`` topic.
    """
    targets = await _markets(state)
    if not targets:
        raise HTTPException(
            503,
            detail={
                "code": "no_markets",
                "message": "BRAIN's settings list no markets to download. Sign in again and retry.",
            },
        )
    return await state.sync.start_all(targets)


@router.get("/sync/runs")
async def list_runs(state: State, limit: int = Query(50, ge=1, le=200)) -> list[dict[str, Any]]:
    return [serialise_run(r) for r in await state.sync.runs(limit)]


@router.get("/sync/runs/{run_id}")
async def get_run(run_id: int, state: State) -> dict[str, Any]:
    run = await state.sync.get_run(run_id)
    if run is None:
        raise HTTPException(404, "No such sync run")
    return serialise_run(run)


@router.post("/sync/runs/{run_id}/cancel")
async def cancel_run(run_id: int, state: State) -> dict[str, bool]:
    return {"cancelled": await state.sync.cancel(run_id)}


@router.get("/sync/last")
async def last_sync(state: State) -> list[dict[str, Any]]:
    """Most recent successful sync per scope — the 'last synced' display."""
    return await state.sync.last_sync_by_tuple()


@router.get("/pyramids")
async def pyramids(state: State) -> dict[str, Any]:
    """Every pyramid: its multiplier, this quarter's alpha count, and download state."""
    return await pyramid_grid(state.endpoints, state.catalog)


@router.get("/scopes")
async def scopes(state: State) -> list[dict[str, Any]]:
    """Which scopes hold data locally, and how much."""
    return await state.queries.synced_tuples()


# --- reading --------------------------------------------------------------


@router.get("/counts")
async def counts(scope: Scope, state: State) -> dict[str, int]:
    """Datasets / categories / subcategories / fields for one scope."""
    return await state.queries.counts(scope)


@router.get("/stats")
async def stats(scope: Scope, state: State) -> dict[str, Any]:
    """Value ranges, so filter controls can bound themselves to real data."""
    return await state.queries.stats(scope)


@router.get("/tree")
async def tree(scope: Scope, state: State) -> list[dict[str, Any]]:
    """Category -> subcategory hierarchy with dataset and field counts."""
    return await state.queries.category_tree(scope)


#: Module-level singleton so the default is not rebuilt per request.
DEFAULT_FIELD_FILTER = FieldFilter()


@router.post("/fields")
async def fields(
    scope: Scope,
    state: State,
    filters: Annotated[FieldFilter, Body()] = DEFAULT_FIELD_FILTER,
) -> dict[str, Any]:
    """Filtered, sorted, paginated data fields.

    POST rather than GET because the filter set is large and structured; the operation
    is still a pure read.
    """
    return await state.queries.fields(scope, filters)


@router.post("/facets")
async def facets(
    scope: Scope,
    state: State,
    filters: Annotated[FieldFilter, Body()] = DEFAULT_FIELD_FILTER,
) -> dict[str, list[dict[str, Any]]]:
    """Categories / subcategories / datasets / types with counts under the other filters."""
    return await state.queries.facets(scope, filters)


@router.get("/fields/{field_id}")
async def field_detail(field_id: str, scope: Scope, state: State) -> dict[str, Any]:
    row = await state.queries.field(scope, field_id)
    if row is None:
        raise HTTPException(404, f"{field_id} is not in the catalog for {scope.label}")
    return row


@router.get("/fields/{field_id}/availability")
async def field_availability(field_id: str, state: State) -> list[dict[str, Any]]:
    """Every scope this field exists in.

    Not every field is available in every region/delay/universe, so this is the check
    that stops a template being expanded into simulations that cannot run.
    """
    return await state.queries.field_availability(field_id)


@router.get("/datasets")
async def datasets(scope: Scope, state: State, search: str | None = None) -> list[dict[str, Any]]:
    return await state.queries.datasets(scope, search)


class CoverageRequest(BaseModel):
    field_ids: list[str]


@router.post("/coverage-matrix")
async def coverage_matrix(payload: CoverageRequest, state: State) -> list[dict[str, Any]]:
    """Availability grid for a set of fields across every synced scope."""
    return await state.queries.coverage_matrix(payload.field_ids)

File: frontend/src/api/catalog.ts
/**
 * The local data-field catalog (Data Explorer). Rows from DuckDB stay snake_case; request
 * bodies are snake_case. `/counts /stats /facets /tree /fields /datasets` read the scope
 * from the query string with `instrumentType`.
 */

import { http, qs } from './http'
import type { Scope, SyncRun } from './types'

/** A market a full sync covers (`GET /api/catalog/markets`). */
export interface Market {
  instrumentType: string
  region: string
  delay: number
  universe: string
}

export interface CatalogSyncRequest {
  region: string
  delay: number
  universe: string
  instrument_type?: string
  restart?: boolean
}

export interface CatalogScopeRow {
  instrument_type: string
  region: string
  delay: number
  universe: string
  fields: number
  synced_at: string | null
}

export interface CatalogCounts {
  fields: number
  datasets: number
  categories: number
  subcategories: number
}

export interface CatalogStats {
  coverage_min: number | null
  coverage_max: number | null
  coverage_median: number | null
  alpha_count_max: number | null
  user_count_max: number | null
  pyramid_multiplier_max: number | null
}

export interface CatalogFacets {
  categories: { id: string; name: string; n: number }[]
  subcategories: { id: string; name: string; n: number; category_id: string | null }[]
  datasets: { id: string; n: number; category_id: string | null; subcategory_id: string | null }[]
  types: { id: string; n: number }[]
}

export interface CategoryNode {
  id: string
  name: string
  datasets: number
  fields: number
  subcategories: { id: string; name: string; datasets: number; fields: number }[]
}

/** Unknown keys sort by alpha_count on the backend. */
export type FieldSortKey =
  | 'field_id'
  | 'dataset_id'
  | 'category_id'
  | 'coverage'
  | 'user_count'
  | 'alpha_count'
  | 'pyramid_multiplier'
  | 'field_type'

export interface FieldFilter {
  search?: string | null
  dataset_ids?: string[]
  category_ids?: string[]
  subcategory_ids?: string[]
  field_types?: string[]
  coverage_min?: number | null
  coverage_max?: number | null
  alpha_count_min?: number | null
  alpha_count_max?: number | null
  user_count_min?: number | null
  user_count_max?: number | null
  pyramid_multiplier_min?: number | null
  sort_by?: FieldSortKey
  sort_desc?: boolean
  limit?: number
  offset?: number
}

export interface DataFieldRow {
  field_id: string
  dataset_id: string | null
  category_id: string | null
  category_name: string | null
  subcategory_id: string | null
  subcategory_name: string | null
  description: string | null
  field_type: string | null
  coverage: number | null
  user_count: number | null
  alpha_count: number | null
  pyramid_multiplier: number | null
  /** JSON array text. */
  themes: string | null
}

export interface FieldPage {
  total: number
  limit: number
  offset: number
  results: DataFieldRow[]
}

export interface DataFieldDetail extends DataFieldRow {
  instrument_type: string
  region: string
  delay: number
  universe: string
  synced_at: string | null
}

export interface FieldAvailabilityRow {
  instrument_type: string
  region: string
  delay: number
  universe: string
  coverage: number | null
  alpha_count: number | null
  field_type: string | null
}

export interface DatasetRow {
  dataset_id: string
  name: string | null
  description: string | null
  category_id: string | null
  category_name: string | null
  subcategory_id: string | null
  subcategory_name: string | null
  coverage: number | null
  value_score: number | null
  user_count: number | null
  alpha_count: number | null
  field_count: number | null
  pyramid_multiplier: number | null
}

export interface PyramidCell {
  categoryId: string
  region: string
  delay: number
  multiplier: number | null
  alphaCount: number
  /** 3+ alphas this quarter. */
  lit: boolean
  synced: boolean
}

export interface PyramidGridData {
  columns: { region: string; delay: number }[]
  categories: { id: string; name: string }[]
  cells: PyramidCell[]
  /** Calendar quarter in platform time; `end` is the next quarter's first day. */
  quarter: { start: string; end: string; today: string }
}

export interface CoverageMatrixRow {
  field_id: string
  region: string
  delay: number
  universe: string
  instrument_type: string
  coverage: number | null
}

const B = '/api/catalog'

/** Scope in the query string, camelCase instrumentType. */
const scopeQs = (s: Scope, extra: Record<string, string | number | undefined> = {}) =>
  qs({ region: s.region, delay: s.delay, universe: s.universe, instrumentType: s.instrumentType, ...extra })

export const catalog = {
  scopes: () => http.get<CatalogScopeRow[]>(`${B}/scopes`),
  sync: (body: CatalogSyncRequest) => http.post<SyncRun>(`${B}/sync`, body),
  syncAll: () => http.post<SyncRun>(`${B}/sync-all`),
  markets: () => http.get<Market[]>(`${B}/markets`),
  runs: (limit = 20) => http.get<SyncRun[]>(`${B}/sync/runs${qs({ limit })}`),
  run: (id: number) => http.get<SyncRun>(`${B}/sync/runs/${id}`),
  cancel: (id: number) => http.post<{ cancelled: boolean }>(`${B}/sync/runs/${id}/cancel`),
  last: () => http.get<SyncRun[]>(`${B}/sync/last`),

  counts: (s: Scope) => http.get<CatalogCounts>(`${B}/counts${scopeQs(s)}`),
  stats: (s: Scope) => http.get<CatalogStats>(`${B}/stats${scopeQs(s)}`),
  /** Counts under every other active filter; each facet ignores its own selection. */
  facets: (s: Scope, filter: FieldFilter) => http.post<CatalogFacets>(`${B}/facets${scopeQs(s)}`, filter),
  tree: (s: Scope) => http.get<CategoryNode[]>(`${B}/tree${scopeQs(s)}`),
  fields: (s: Scope, filter: FieldFilter) => http.post<FieldPage>(`${B}/fields${scopeQs(s)}`, filter),
  field: (s: Scope, id: string) => http.get<DataFieldDetail>(`${B}/fields/${encodeURIComponent(id)}${scopeQs(s)}`),
  availability: (id: string) => http.get<FieldAvailabilityRow[]>(`${B}/fields/${encodeURIComponent(id)}/availability`),
  datasets: (s: Scope, search?: string) => http.get<DatasetRow[]>(`${B}/datasets${scopeQs(s, { search })}`),

  pyramids: () => http.get<PyramidGridData>(`${B}/pyramids`),
}
